lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/group/CachingGroupProviderModule.java

```diff
@@ -0,0 +1,75 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.base.group;
+
+import com.google.inject.Binder;
+import com.google.inject.BindingAnnotation;
+import com.google.inject.Key;
+import com.google.inject.Scopes;
+import io.airlift.configuration.AbstractConfigurationAwareModule;
+import io.trino.spi.security.GroupProvider;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import static io.airlift.configuration.ConditionalModule.conditionalModule;
+import static io.airlift.configuration.ConfigBinder.configBinder;
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * If added to the list of {@link com.google.inject.Module}s used in initialization of a Guice context in a
+ * {@link io.trino.spi.security.GroupProviderFactory}, it will (almost) automatically add caching capability to the
+ * group provider. Requirements:
+ * <ul>
+ *     <li>The {@link GroupProvider} available in the Guice context must be bound annotated with
+ *     {@link ForCachingGroupProvider} binding annotation</li>
+ * </ul>
+ * The module will make the following configuration options available (to be set in {@code etc/group-provider.properties}:
+ * <ul>
+ *     <li>{@code group-provider.cache.enabled} - the toggle to enable or disable caching</li>
+ *     <li>{@code group-provider.cache.ttl} - determines how long group information will be cached for each user</li>
+ *     <li>{@code group-provider.cache.maximum-size} - maximum number of users for which groups are stored in the cache</li>
+ * </ul>
+ */
+public class CachingGroupProviderModule
+        extends AbstractConfigurationAwareModule
+{
+    @Override
+    protected void setup(Binder binder)
+    {
+        configBinder(binder).bindConfig(GroupProviderConfig.class);
+        install(conditionalModule(
+                GroupProviderConfig.class,
+                GroupProviderConfig::isCachingEnabled,
+                innerBinder -> {
+                    configBinder(innerBinder).bindConfig(CachingGroupProviderConfig.class);
+                    innerBinder.bind(CachingGroupProvider.class).in(Scopes.SINGLETON);
+                    innerBinder.bind(GroupProvider.class)
+                            .to(CachingGroupProvider.class).in(Scopes.SINGLETON);
+                },
+                innerBinder -> innerBinder.bind(GroupProvider.class)
+                        .to(Key.get(GroupProvider.class, ForCachingGroupProvider.class))
+                        .in(Scopes.SINGLETON)));
+    }
+
+    @Retention(RUNTIME)
+    @Target({FIELD, PARAMETER, METHOD})
+    @BindingAnnotation
+    public @interface ForCachingGroupProvider
```

Comment:
nit `public @interface ForCachingGroupProvider {}` in one line

Reply:
I haven't seen this formatting in any of the other binding annotations :)